bundler/lib/bundler/dep_proxy.rb

```diff
@@ -4,19 +4,18 @@ module Bundler
   class DepProxy
     attr_reader :__platform, :dep
 
+    @proxies = {}
+
+    def self.get_proxy(dep, platform)
+      @proxies[[dep, platform].hash] ||= new(dep, platform).freeze
```

Comment:
I think the key here should be `[dep, platform]`, instead of that array's hash, otherwise hash collisions would lead to the wrong instance being returned

Reply:
I mean, that's simpler and I can update it, but I don't understand why the current version could cause issues.

`Object` docs say:

> The hash value for an object may not be identical across invocations or implementations of Ruby. If you need a stable identifier across Ruby invocations and implementations you will need to generate one with a custom method.

That means it _is_ stable across the same invocation, so `[dep, platform].hash` will always return the same thing, which is what we want here, right?

Reply:
I think the danger is not that the value might change, but that _different_ objects could have the same hash. Hash collisions have definitely been the cause of horrible Bundler bugs in the past, on 32-bit machines where completely different values hashed to the same hash. 😬

Reply:
Oh I get it now! Sure, better stay on the safe side, a bug there would be terrible :grimacing:.

Reply:
Nice catch @segiddins!

Reply:
However, are we sure that using an array as the hash key wouldn't be problematic either? It seems to me that ruby uses `#hash` under the hood anyways so maybe it has the same issue.

Maybe

```diff
diff --git a/bundler/lib/bundler/dep_proxy.rb b/bundler/lib/bundler/dep_proxy.rb
index c7e2b40634..0c7eb8fdb6 100644
--- a/bundler/lib/bundler/dep_proxy.rb
+++ b/bundler/lib/bundler/dep_proxy.rb
@@ -4,10 +4,10 @@ module Bundler
   class DepProxy
     attr_reader :__platform, :dep
 
-    @proxies = {}
+    @proxies = Hash.new{|hash, key| hash[key] = {} }
 
     def self.get_proxy(dep, platform)
-      @proxies[[dep, platform].hash] ||= new(dep, platform).freeze
+      @proxies[dep][platform] ||= new(dep, platform).freeze
     end
 
     def initialize(dep, platform)
```

is better?

Reply:
I think, although I am not 100% sure, that the implementation of `Hash` uses .hash, but does not treat two objects with the same hash as identical. Like most hash tables or hash maps, I believe it uses the hash value to determine a bucket, and then uses object equality checks within that bucket. So if we use `.hash`, I think we are removing the Hash implementation's ability to tell the difference between two objects that have the same hash but are not identical objects. That said, I don't have time to check the MRI implementation right now, so I could be wrong!

Reply:
Still, are we sure that different arrays can have the same `#hash`? It really seems super weird to me, why would ruby do that? Shouldn't it generate a unique value for each array?

Reply:
Ok, I think the docs are pretty clear that this can't happen: https://ruby-doc.org/core-3.0.0/Array.html#method-i-hash. I think there's no issue here.

Reply:
Here's a somewhat reasonable SO discussion on the subject: https://stackoverflow.com/questions/36952150/ruby-internals-and-how-to-guarantee-unique-hash-values. Calculated hash values can always have collisions, because the hash result has less possible values than there are possible objects. You cannot guarantee every possible object will have a unique hash, so you have to use it as an optimization step rather than the final check.

Reply:
(@indirect I missed to read your answer before posting my comments, checking it now)

Reply:
(IIRC, hash collisions caused occasional confusing bugs, but we didn't realize what was happening until someone showed up with a Ruby compiled to run on a 16-bit CPU, where the possible hash values were so few that hash collisions were happening every few seconds that Bundler was running.)

Reply:
I _think_ we are confusing the `#hash` (_method_) ruby uses to determine the key used under the hood when using objects as keys with the hash (function) used by the hash (data structure) implementation to map keys to buckets. The hash (function) can certainly have collisions that should be handled by the implementation, but I don't see why the `#hash` method on an Array should have collisions and not be unique to the array contents.

Reply:
Before going with these alternate implementations, make sure that two *different* Gem::Dependency objects with identical contents really do create the same lookup key. Otherwise something like `to_s` could be used. So we could go with 

```ruby
@proxies["#{dep.to_s}::#{platform.to_s"] ||= new(dep, platform).freeze
```

Assuming both `to_s` functions have all the contents that make deps and platforms unique.

Without the cache hits on creation, this pr will not improve anything.

Reply:
I think `to_s` wouldn't work because we want things like "foo >= 2.0" and  "foo >= 2.0.0" to get the same key. But the other alternatives suggested I think would work equally fine.